File: js/src/rle.ts



class Last {
    val: Rle = null;
    max_poses: { [id: string] : any } = {};
    ret: Rle[] = [];

    constructor() {
    }
    public handle_last() {
        if (null == this.val) {
            return;
        }
        let _last = this.val;
        let max_rles = this.max_poses[_last.part]; 
        if (max_rles == null) {
            max_rles = this.max_poses[_last.part] = [];
        }
        for (let idx in max_rles) {
            let prev = this.ret[idx];
            if (prev.cnt > _last.cnt) {
                console.log(`>>>>> last=${_last}->${idx}->prev=${prev}`);
                _last.max = false;
            } else if (prev.cnt == _last.cnt) {
                // nothing
            } else if (prev.cnt < _last.cnt) {
                console.log(`<<<<< last=${_last}->${idx}->prev=${prev}`);
                //self.ret[idx].max = false;
                prev.max = false;
            }
        }
        //println!("push:{}:{:?}", self.ret.len(), _last);
        max_rles.push(this.ret.length);
        _last.pos = this.ret.length;
        this.ret.push(_last);
    }
}

class Rle {
    public part: any;
    public pos: number;
    public cnt: number;
    public max: boolean;

    constructor(obj: { [id: string]: any }) {
        this.part = obj['part'];
        this.pos = obj['pos'];
        this.cnt = obj['cnt'];
        this.max = obj['max'];
    }

    public toString() {
        return `<Rle@part:${this.part},pos:${this.pos},cnt:${this.cnt},max:${this.max}>`;
    }

    public eq(other: Rle): boolean {
        return this.part == other.part && this.pos == other.pos &&
            this.cnt == other.cnt && this.max == other.max;
    }
    public ne(other: Rle): boolean {
        return !this.eq(other);
    }

    public static code(parts: any[]): Rle[] {
        let last = new Last();
        // println!("code");
        for (let i=0; i < parts.length; ++i) {
            let part = parts[i];
            // println!("part:{}", part);
            if (last.val && last.val.part == part) {
                last.val.cnt += 1;
            } else {
                last.handle_last();
                last.val = new Rle({ part: part, pos: 0, cnt: 1, max: true });
            }
        }
        last.handle_last();
        return last.ret;
    }

}

export default Rle;

File: cpp/src/rle.cpp
#include <iostream>

#include "option.hpp"

#include "rle.hpp"

std::ostream& operator<<(std::ostream &o, const Rle &rle) {
    o << rle.toString();
    return o;
}


class Last {
public:
    Option<Rle> val;
    typedef std::map<size_t, std::vector<size_t>> IdxMap;
    IdxMap max_poses;
    std::vector<Rle> ret;

    Last() {}

    void handle_last() {
        if (this->val.isNone()) {
            return;
        }
        auto &_last = this->val.unwrap();
        // std::cout << "handle_last:" << _last.part << std::endl;
        auto max_rles = this->max_poses.find(_last.part);
        if (max_rles == this->max_poses.end()) {
            max_rles = this->max_poses.insert(IdxMap::value_type(_last.part, std::vector<size_t>())).first;
        }
        // console.log(_last.part, this->max_poses);
        for (auto idx : max_rles->second) {
            // std::cout << "max_rles:" << idx;
            Rle &prev = this->ret[idx];
            if (prev.cnt > _last.cnt) {
                // console.log(`>>>>> last=${_last}->${idx}->prev=${prev}`);
                // std::cout << "_last.max=" << idx << " " << _last << std::endl;
                _last.max = false;
            } else if (prev.cnt == _last.cnt) {
                // nothing
            } else if (prev.cnt < _last.cnt) {
                // console.log(`<<<<< last=${_last}->${idx}->prev=${prev}`);
                // std::cout << "prev.max=" << idx << " " << prev << std::endl;
                prev.max = false;
            }
        }
        //println!("push:{}:{:?}", self.ret.len(), _last);
        max_rles->second.push_back(this->ret.size());
        _last.pos = this->ret.size();
        this->ret.push_back(_last);
    }
};

std::vector<Rle> Rle::code(const std::vector<size_t> &parts) {
    Last last;
    // println!("code");
    for (auto part : parts) {
        // console.log(`part:${part}`);
        if (last.val.isSome() && last.val.unwrap().part == part) {
            last.val.unwrap().cnt += 1;
        } else {
            last.handle_last();
            Rle rle(part, 0, 1, true);
            last.val = Some(rle);
        }
    }
    last.handle_last();
    return last.ret;
}
